File: benchmarks/runtime/programs/c/string-processing.c
static const char line_text[] = "GET /users/42 HTTP/1.1 200 1532";
static const int line_len = (int)(sizeof line_text - 1);

static int count_byte(const char *text, int len, int target) {
    int found = 0;
    for (int i = 0; i < len; i++) {
        if ((unsigned char)text[i] == target) {
            found++;
        }
    }
    return found;
}

static int count_digits(const char *text, int len) {
    int found = 0;
    for (int i = 0; i < len; i++) {
        unsigned char byte = (unsigned char)text[i];
        if (byte >= 48 && byte <= 57) {
            found++;
        }
    }
    return found;
}

static int slice_equals(const char *text, int start, int end, const char *expect) {
    for (int i = start; i < end; i++) {
        if (text[i] != expect[i - start]) {
            return 0;
        }
    }
    return expect[end - start] == '\0';
}

static int score(const char *text, int len) {
    return count_byte(text, len, 32) + count_byte(text, len, 47) +
           count_digits(text, len) + slice_equals(text, 0, 3, "GET") +
           slice_equals(text, 14, 22, "HTTP/1.1");
}

int main(void) {
    int total = 0;
    for (int round = 0; round < 200; round++) {
        total += score(line_text, line_len);
    }
    return total;
}

File: benchmarks/runtime/programs/go/string-processing.go
// string-processing — the equivalent-semantics Go peer for the tuonelang
// `string-processing` workload (ADR-0006's borrowed `Str`). Byte-level scanning
// over a fixed request-log line, done the same way: count spaces (32) and
// slashes (47), count ASCII digits, and compare two fixed slices against "GET"
// and "HTTP/1.1". Go's []byte view of a string is the peer to tuonelang's byte
// slicing over `Str`. Per round: 4 spaces + 3 slashes + 11 digits + 1 + 1 = 20;
// 200 rounds × 20 = 4000; observable exit byte = 4000 & 0xff = 160.
package main

import "os"

var lineText = []byte("GET /users/42 HTTP/1.1 200 1532")

func countByte(text []byte, target byte) int {
	found := 0
	for i := 0; i < len(text); i++ {
		if text[i] == target {
			found++
		}
	}
	return found
}

func countDigits(text []byte) int {
	found := 0
	for i := 0; i < len(text); i++ {
		b := text[i]
		if b >= 48 && b <= 57 {
			found++
		}
	}
	return found
}

func sliceEquals(text []byte, start, end int, expect string) int {
	if end-start != len(expect) {
		return 0
	}
	for i := start; i < end; i++ {
		if text[i] != expect[i-start] {
			return 0
		}
	}
	return 1
}

func score(text []byte) int {
	return countByte(text, 32) + countByte(text, 47) +
		countDigits(text) + sliceEquals(text, 0, 3, "GET") +
		sliceEquals(text, 14, 22, "HTTP/1.1")
}

func main() {
	total := 0
	for round := 0; round < 200; round++ {
		total += score(lineText)
	}
	os.Exit(total)
}
